### DISCUSSION TITLE
GRE Key as Policy Selector

### DISCUSSION THREAD

**Discussion by andrew-hoff** [strongswan/strongswan#314] [Q&A] [answered]
Hi all. I have a setup where I would like to have multiple peers behind the same NAT communicate with a cloud node. This means that from the cloud node's perspective, the nodes behind the NAT will have the same local/remote pair. I would like to use a route-based VPN to connect as well. VTI does not seem to be a viable option because netlink rejects two VTI interfaces with the same local/remote pair, even if they have different marks. XFRM is also not an option for other reasons not worth going into here.

My next thought is to use GRE. After some initial testing, I found that you can create two GRE interfaces with the same local/remote pair so long as the keys are distinct. However, given that both interfaces have the same local/remote pair, my policy will need to use the key to identify the appropriate security association for packets routed through each interface.

Ideally, I can use strongswan to configure the UPSPEC field with the appropriate GRE key, as specified [here](https://man7.org/linux/man-pages/man8/ip-xfrm.8.html). However, so far I don't see a good way to do that in the documentation. My backup plan is to just create an iptables rule to mark packets that go through each interface with the corresponding GRE key and use mark_in/mark_out on the SA. Is that my best option, or is there a better way to do this in strongswan?

**Comment by Thermi** [comment, marked as answer]
remote_ts and local_ts = %dynamic[gre/x]

**Comment by andrew-hoff** [comment]
Ah, you can use the '/x' as the GRE rather than the port?

**Reply by Thermi** [reply]
Yes, as the GRE key.E.g. [gre/1.1.1.1] AFAIK (check source code for details).

**Reply by andrew-hoff** [reply]
Great, I will dig into the source code and try this out. Thank you!

**Reply by andrew-hoff** [reply]
I dug through the code and it looks like this is _not_ the case. In particular, [this](https://github.com/strongswan/strongswan/blob/master/src/libcharon/plugins/vici/vici_config.c#L685) section looks as though all fields after the `/` are simply converted to ports.

That having been said, I'm not an expert in this codebase!

**Reply by Thermi** [reply]
You can, but the x.x.x.x notation does not work (just tried it):
```
onnections {
        gre-test {
                children {
                        gre-test {
                                local_ts = dynamic[gre/0x152], dynamic[gre/123]
                        }
                }
        }

}
```
-> 
```
gre-test: IKEv1/2, no reauthentication, rekeying every 14400s
  local:  %any
  remote: %any
  local unspecified authentication:
  remote unspecified authentication:
  gre-test: TUNNEL, rekeying every 3600s
    local:  dynamic[gre/338] dynamic[gre/123]
    remote: dynamic
```

I do not know if it works effectively though.

**Reply by andrew-hoff** [reply]
Sure, but doesn't that refer to the port, not the key? My biggest hunch for this was that the field specified is limited to two bytes as can be seen [here](https://github.com/strongswan/strongswan/blob/master/src/libcharon/plugins/vici/vici_config.c#L728). GRE keys can be up to 4 bytes.

**Reply by Thermi** [reply]
Yes, you're right in that regard.

**Reply by Thermi** [reply]
WELL, I just tested it and it works. Kind of.

`ip x p`:
```
src 192.168.178.20/32 dst 192.0.0.5/32 proto gre key 22151168 
        dir out priority 366912 ptype main 
        tmpl src 192.168.178.20 dst 192.0.0.5
                proto esp reqid 1 mode tunnel
src 192.0.0.5/32 dst 192.168.178.20/32 proto gre key 338 
        dir fwd priority 366912 ptype main 
        tmpl src 192.0.0.5 dst 192.168.178.20
                proto esp reqid 1 mode tunnel
src 192.0.0.5/32 dst 192.168.178.20/32 proto gre key 338 
        dir in priority 366912 ptype main 
        tmpl src 192.0.0.5 dst 192.168.178.20
                proto esp reqid 1 mode tunnel
src 192.168.178.20/32 dst 192.0.0.5/32 proto gre key 8060928 
        dir out priority 366912 ptype main 
        tmpl src 192.168.178.20 dst 192.0.0.5
                proto esp reqid 1 mode tunnel
src 192.0.0.5/32 dst 192.168.178.20/32 proto gre key 123 
        dir fwd priority 366912 ptype main 
        tmpl src 192.0.0.5 dst 192.168.178.20
                proto esp reqid 1 mode tunnel
src 192.0.0.5/32 dst 192.168.178.20/32 proto gre key 123 
        dir in priority 366912 ptype main 
        tmpl src 192.0.0.5 dst 192.168.178.20
                proto esp reqid 1 mode tunnel
```

That happens if you install a policy with the config I pasted earlier.

**Reply by andrew-hoff** [reply]
Got it working! Thank you @Thermi !

**Reply by Thermi** [reply]
yw. :)
